### DISCUSSION TITLE
Alpine.js x-bind not applying after an HTMX swap

### DISCUSSION THREAD

**Discussion by yaakovLowenstein** [alpinejs/alpine#3985] [7. Bugs (archived)] [answered]
Hi. I'm not sure this officially counts as a bug but I ran into the same issue as this post :

https://stackoverflow.com/questions/76891618/alpine-js-x-bind-not-applying-after-an-htmx-swap

The issue is that when swapping an element with htmx the x-binding was not working after the swap. (The state was not being swapped so that was not the issue.)

Someone in the comments suggested removing the id from the element and that worked! 

I have no idea why that worked and it doesn't sound like the one who wrote that in the StackOverflow post understood why either.

Here is the code I am using (this is without the id):
```html
<div 
                x-data="{value:{{item.pk}}}"
                class=" mock-checkbox"
                :class="   {'checked':$store.checked.values.includes($data.value) }"
                @click="$el.classList.contains('checked') ?  $store.checked.remove($data.value) : $store.checked.add($data.value)">
                <span class='mock-checkbox-check'>&#10003</span>
</div>
```

**Comment by ekwoka** [comment]
:thinking: That's curious. the id shouldn't really matter unless htmx is doing something strange with them...

**Comment by SimoTod** [comment]
Are you using https://htmx.org/extensions/alpine-morph/?

**Reply by yaakovLowenstein** [reply]
No, I saw it but didn't think I needed it for my case because the state lives outside of the element  being swapped.

I even checked the value of the state after the swap and it had the value I expected but the x bind was not triggering off of that state

**Reply by SimoTod** [reply]
I suspect it's not going to work because Alpine didn't see an attribute change but probably in the underlying DOM, that element is a different one so the reactive effect is bound to an element that doesn't exist anymore. Likely some incompatibility between Alpine and htmx swapping mechanism. I would try that plugin.

**Comment by ekwoka** [comment]
Oooh, yes, are you using alpine morph plugin?

I can't imagine this issue would happen without using that since it should just be seen as totally different elements...

**Reply by yaakovLowenstein** [reply]
I'm not using it. You think that it would only happen if I am using it?

**Reply by ekwoka** [reply]
Since the elements with morph would try to "merge" the information and some can break. but htmx just drops them entirely, which SHOULD totally reset it as if it was a fresh render...

**Comment by yaakovLowenstein** [comment]
Don't know if this helps at all but this is a gif of what I'm working on. And it shows even stranger behavior as well. ( I didn't know how to  describe what's happening so I figured a gif would be best)
![checkbox_behavior](https://github.com/alpinejs/alpine/assets/50084761/038a4cab-25d5-470f-be56-e4768e367ae3)

**Reply by yaakovLowenstein** [reply]
you can't see the mouse which is annoying🤦.
1. I click the first 2 checkboxes
2. I filter the table by Nov 23 ( the table is swapped via htmx) where you can see that nothing is checked (the first checkbox should be)
3.  I reset filters and only the last 2 are checked. 
4. The first checkbox is unclickable and when I click a new checkbox it also checks the first (this happens because the value of the first is still in the store, but it's strange that it's not checked and that it can't be clicked but checking a new one triggers that one to be checked)

**Reply by ekwoka** [reply]
That is really strange...

**Comment by ekwoka** [comment]
Oh, your conditions are just strange and causing issues.

You should just be using `x-model` on the checkbox, not the :class :checked etc.

**Reply by yaakovLowenstein** [reply]
Hmm.. thanks for the suggestion 

1. It is all just really strange but it's also odd that removing the id fixes Al of this

2. How do you set a class with x-model?

3. In this case I don't think I can use it because I'm not using real checkboxes, I had other issues with those which I won't get into here

**Reply by SimoTod** [reply]
I don't know the internal for htmx but if I had to guess, I would say that they use the id to track the elements. If there's an id, they somehow morph the element and if there isn't, they remove the old element and add a new one (which Alpine can track easily). Now that I'm thinking about that, the issue is probably that htmx morphs the element and removes the class attribute (because the new element doesn't have it), however Alpine doesn't recompute the binding unless something has changed.

**Reply by ekwoka** [reply]
For the class, you don't need the class at all.

the element will already be selectable with the :checked pseudo-selector.

Should really just use real checkboxes or it will be an accessibility nightmare. Like right now, you can only use a mouse to click them...

**Reply by yaakovLowenstein** [reply]
Thank you for pointing me towards x-model I had read about it but didn't fully grasp it.

I got it to work with checkboxes using x-data and x-model.

I had to put the x-data though on an element out of the htmx that was being swapped even when using the morph plugin from htmx. I'm not sure why that is because I thought the whole point was to keep the state.

I have been working htmx for awhile and am new to alpinejs. I still do not understand a lot of it and am having trouble grasping why some things work and some don't, but overall it is very cool and does work nicely with htmx.

**Reply by yaakovLowenstein** [reply]
@SimoTod is there somewhere I could read to fully understand the way alpine watches and knows when something has changed.

You mentioned that without an id it would be a new element which alpine can watch easily vs if it's morphed in. How come that is?

**Reply by SimoTod** [reply]
Only the source code as it's not something the majority of developers really cares about: https://github.com/alpinejs/alpine/blob/main/packages/alpinejs/src/mutation.js

https://developer.mozilla.org/en-US/docs/Web/API/MutationObserver

The summary is:
If you add an element, the mutation observer runs every new Alpine directive.
If you remove an element, the mutation observer destroys all effects linked to its Alpine directives.
If you add an attribute and it's an Alpine directive, the mutation observer runs that directive.
If you remove an attribute and it's an Alpine directive, the mutation observer destroys all effects linked to that directive.
If you add/remove a normal attribute, the mutation observer does nothing.

**Reply by yaakovLowenstein** [reply]
Thank you!

**Comment by shimikano** [comment]
Hello

I ran into the same issue and could reproduce it in a Codepen: https://codepen.io/shimikano/pen/MWREmPO

The component in question is:

```html
<div x-data="{ valid: true }">
  <label for="foo" x-text="'The border should be ' + (valid ? 'green' : 'red') + ': '"></label>
  <input id="foo" name="foo" value="bar" :class="valid ? 'valid' : 'invalid'">
</div>
```

When htmx swaps in the same markup with `valid` set to `false`, the `:class` binding on the `<input>` is applied but quickly cleared (border flashes red). Removing the `id` attribute from the `input` solves the issue.

Does anyone have some insight as to how and where (alpine or htmx) this can be tackled?

Using plain htmx and alpinejs, without any plugins.

Thank you.

**Reply by SimoTod** [reply]
I think you need https://htmx.org/extensions/alpine-morph/ if you want htmx and alpinejs to play together nicely otherwise htmx will mess up with the state of your alpine components

**Reply by shimikano** [reply, marked as answer]
Thank you for your reply @SimoTod.

After some more digging, I understand now that the culprit is the default htmx swap/settle algorithm for elements with an `id` (documented [here](https://htmx.org/docs/#details)). There is also a corresponding [htmx issue](https://github.com/bigskysoftware/htmx/issues/412).

Possible workarounds include:

  - Remove the `id` attribute.
  - Set the htmx settle delay to zero (`settle:0`).
  - Use an alternative swap algorithm such as the alpine morph extension you mentioned. This also preserves the alpine state across swaps, which may or may not be what you want.
  - Remove the attributes in question from `htmx.config.attributesToSettle`.

I updated the Codepen, demonstrating 2 workarounds.

**Comment by 1cg** [comment]
Hey folks, I am working on the next version of htmx and I'd like to make this "just work".  Is there anything we can do on our side to signal to alpine that the content has been updated?

**Comment by 1cg** [comment]
Heya folks, just wanted to give this group an update: we (the htmx team) did a deep dive on the Alpine codebase and identified two areas where we can dramatically improve htmx compatibility w/alpine and one change alpine could make to improve compatibility with us.  We are gonna put out a compatibility extension soon (hopefully for both v2 and the still-in-alpha-4) in the next week or so.
